### PR TITLE
[core] Fix #10404: Fix interaction of TThreadExecutor and TSeq with specified beginning/range

### LINKED ISSUE #10404: Broken interaction between `TThreadExecutor` and `TSeq` with specified beginning/range
# Problem:

`TThreadExecutor` expects that `TSeq`-s always start with 0 and have step 1. Otherwise, it will give wrong result for `Foreach`, `MapReduce`, `Map` of the `TThreadExecutor`.

With current master, the following snipet produces 0 and depending on compiler can also give `munmap_chunk(): invalid pointer` followed by `SIGABRT`:
```C++
#include "ROOT/TThreadExecutor.hxx"
#include "ROOT/TSeq.hxx"
#include <iostream>

int main(){
   auto func = [&](long int x)->long int { return x; };
   auto redfunc = [](const std::vector<long int> &v){return std::accumulate(v.begin(), v.end(), 0l);};
   ROOT::TThreadExecutor t(1);
   std::cout << t.MapReduce(func, ROOT::TSeq<long int>(2, 4), redfunc) << '\n'; // expected result: 2 + 3 = 5
   return 0;
}
```
Moreover, when range is specified there is seg fault produced.

# Partial Solution:

Looking with `gdb` I saw that in the reduction step, there is a call inside `TThreadExecutor.hxx:270`:
```C++
      unsigned start = *args.begin();
      unsigned end = *args.end();
      unsigned seqStep = args.step();

      using retType = decltype(func(start));
      std::vector<retType> reslist(args.size());
      
      auto lambda = [&](unsigned int i)
      {
         reslist[i] = func(i);
      };
      ParallelFor(start, end, seqStep, lambda);
``` 
This assumes that the TSeq always starts with 0 and has step size 1. 

To properly handle TSeq for MapReduce need to have:
```C++
      auto lambda = [&](unsigned int i)
      {
         reslist[i] = func(args[i]);
      };
      ParallelFor(0U, args.size(), 1, lambda); // this calls tbb::parallel_for(start, end, step, f)
      // tbb::parallel_for expects as args: 1.starting idx, 2. end index, 3. step size, 4. func
      // docs: https://spec.oneapi.io/versions/1.1-rev-1/elements/oneTBB/source/algorithms/functions/parallel_for_func.html
```

### PR DESCRIPTION
**Pull request by ikabadzhov**
* Modified the logic of MapReduce and Map of the TThreadExecutor which were
initially broken for the TThreadExecutor;
* Added corresponding tests.

Fix https://github.com/root-project/root/issues/10404

### CODE REVIEW

**Review by eguiraud — APPROVED** [review]
LGTM, thanks!

**Comment by ikabadzhov** [comment]
> I think the failures are real

Unfortunately, they indeed are. I realized where the problem is coming from. I still need to figure out how to solve it though.

**Comment by ikabadzhov** [comment]
@phsft-bot build just on windows10/cxx14

**Comment by ikabadzhov** [comment]
@phsft-bot build just on windows10/cxx14

File: core/imt/test/testRTaskArena.cxx [hunk 1]
```diff
@@ -246,4 +246,35 @@ TEST(TThreadExecutor, ThreadSafety) {
...
+
+   // Map on TSeq with end specified only
+   EXPECT_EQ(redfunc(ttex.Map(func, ROOT::TSeq<long int>(5))), 10);
+
+   // Map on TSeq with begin and end specified only
+   EXPECT_EQ(redfunc(ttex.Map(func, ROOT::TSeq<long int>(2, 5))), 9);
+
+   // Map on increasing and decreasing TSeq with begin, end and step specified
+   EXPECT_EQ(redfunc(ttex.Map(func, ROOT::TSeq<long int>(2, 5, 2))), 6);
+   EXPECT_EQ(redfunc(ttex.Map(func, ROOT::TSeq<long int>(5, 2, -2))), 8);
```
Comment (eguiraud) [inline core/imt/test/testRTaskArena.cxx:277]: why not just `int`? then you can just write `TSeqI` which reduces noise a bit

**Comment by eguiraud** [comment]
@phsft-bot build please

**Review by eguiraud — APPROVED** [review]
LGTM if jenkins agrees (one last time), and the new logic also reads better than the old one.

**Pull request closed without merging**